File: NextSpring/src/main/java/com/study/nextspring/global/security/JwtAuthenticationFilter.java
package com.study.nextspring.global.security;

import com.study.nextspring.domain.member.service.MemberService;
import jakarta.servlet.FilterChain;
import jakarta.servlet.http.*;
import lombok.*;
import org.springframework.stereotype.Component;
import org.springframework.web.filter.OncePerRequestFilter;
import com.study.nextspring.global.httpsdata.*;

import java.util.List;

@Component
@RequiredArgsConstructor
public class JwtAuthenticationFilter extends OncePerRequestFilter {
    private final ReqData rq;
    private final MemberService memberService;

    @Override
    @SneakyThrows
    protected void doFilterInternal(HttpServletRequest request, HttpServletResponse response, FilterChain filterChain) {
        String accessToken_ = rq.getCookieValue("accessToken", "");
        System.out.println("accessToken from cookie: " + accessToken_);

        if (!request.getRequestURI().startsWith("/api/")) {
            filterChain.doFilter(request, response);
            return;
        }

        // 로그인, 회원가입, 로그아웃 API는 제외
        if (List.of("/api/v1/members/login", "/api/v1/members/join", "/api/v1/members/logout").contains(request.getRequestURI())) {
            filterChain.doFilter(request, response);
            return;
        }

        // Authorization 헤더에서 Bearer 토큰 처리
        String bearerToken = rq.getHeader("Authorization", null);
        if (bearerToken == null || !bearerToken.startsWith("Bearer ")) {
            // 헤더에 토큰이 없으면 쿠키에서 가져오도록 처리
            accessToken_ = rq.getCookieValue("accessToken", "");
            if (accessToken_.isBlank()) {
                filterChain.doFilter(request, response);
                return;  // accessToken이 없으면 필터를 계속해서 실행하지 않음
            }
        } else {
            // Bearer 토큰이 있는 경우
            String tokensStr = bearerToken.substring("Bearer ".length());
            String[] tokens = tokensStr.split(" ", 2);
            String refreshToken = tokens[0];
            String accessToken = tokens.length == 2 ? tokens[1] : "";

            if (!accessToken.isBlank()) {
                // accessToken 유효성 체크
                if (!memberService.validateToken(accessToken)) {
                    System.out.println("Access token expired, refreshing...");
                    RespData<String> rs = memberService.refreshAccessToken(refreshToken);
                    accessToken = rs.getData();
                    rq.setHeader("Authorization", "Bearer " + refreshToken + " " + accessToken);
                }

                SecurityUser securityUser = memberService.getUserFromAccessToken(accessToken);
                System.out.println("Logged in user: " + securityUser.getUsername());
                rq.setLogin(securityUser);
            }
        }

        filterChain.doFilter(request, response);
    }

}

File: k_rstudy/src/main/kotlin/com/krstudy/kapi/global/Security/JwtAuthenticationFilter.kt
package com.krstudy.kapi.global.Security

import com.krstudy.kapi.domain.member.service.MemberService
import com.krstudy.kapi.global.https.ReqData
import com.krstudy.kapi.global.https.RespData
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import lombok.RequiredArgsConstructor
import org.springframework.stereotype.Component
import org.springframework.web.filter.OncePerRequestFilter

@Component
@RequiredArgsConstructor
class JwtAuthenticationFilter(
    private val rq: ReqData,
    private val memberService: MemberService
) : OncePerRequestFilter() {

    @Throws(Exception::class)
    override fun doFilterInternal(request: HttpServletRequest, response: HttpServletResponse, filterChain: FilterChain) {
        if (!request.requestURI.startsWith("/api/")) {
            filterChain.doFilter(request, response)
            return
        }

        val bearerToken = rq.getHeader("Authorization","") ?: return filterChain.doFilter(request, response)

        val tokensStr = bearerToken.substring("Bearer ".length).takeIf { it.isNotBlank() } ?: return
        val tokens = tokensStr.split(" ", limit = 2)
        val refreshToken: String = tokens[0]
        var accessToken: String = tokens.getOrElse(1) { "" }

        // 엑세스 토큰이 존재하면
        if (accessToken.isNotBlank()) {
            // 유효성 체크하여 만료되었으면 리프레시 토큰으로 새로운 엑세스 토큰을 발급받고 응답헤더에 추가
            if (!memberService.validateToken(accessToken)) {
                val rs: RespData<String> = memberService.refreshAccessToken(refreshToken)
                accessToken = rs.data ?: return filterChain.doFilter(request, response) // 데이터가 null일 경우 처리
                rq.setHeader("Authorization", "Bearer $refreshToken $accessToken")
            }

            val securityUser = memberService.getUserFromAccessToken(accessToken)
            rq.setLogin(securityUser)
        } else {
            // 토큰이 쿠키로 들어온 경우
            var accessToken = rq.getCookieValue("accessToken", "") ?: ""

            // 엑세스 토큰이 존재하면
            if (accessToken.isNotBlank()) {
                // 유효성 체크하여 만료되었으면 리프레시 토큰으로 새로운 엑세스 토큰을 발급받고 응답쿠키에 추가
                if (!memberService.validateToken(accessToken)) {
                    val refreshToken = rq.getCookieValue("refreshToken", "") ?: ""

                    val rs: RespData<String> = memberService.refreshAccessToken(refreshToken)
                    accessToken = rs.data ?: return filterChain.doFilter(request, response) // 데이터가 null일 경우 처리
                    rq.setCrossDomainCookie("accessToken", accessToken)
                }

                val securityUser = memberService.getUserFromAccessToken(accessToken)
                rq.setLogin(securityUser)
            }
        }

        filterChain.doFilter(request, response)
    }
}
